### PR TITLE
[reflection] Implement RuntimeType.IsByRefLike

### LINKED ISSUE #11086: Type.IsByRefLike throws NotSupportedException
## Steps to Reproduce

```csharp
int test = 10;
Console.WriteLine (test.GetType ().IsByRefLike);
```

### Current Behavior

```
   System.NotSupportedException : Derived classes must provide an implementation.
  at System.Type.get_IsByRefLike () [0x00000] in /Users/alexander/dev/mono/mcs/class/corlib/ReferenceSources/Type.cs:42
```

### Expected Behavior

No exception and `false` for types like int.

## On which platforms did you notice this

[ X ] macOS
[ X ] Linux
[ X ] Windows

**Version Used**: master

### PR DESCRIPTION
**Pull request by lambdageek**
Fixes #11086

- [x] Add a byreflike bit to `MonoClass`
- [x] Set bit when we create a class from metadata; or when we create a ginst
- runtime checks
   - [x] no byreflike arrays
   - [x] byreflike only allowed as a field iff it's an instance field and contained in another byreflike struct
   - [x] no boxing
   - [x] can't instantiate generic params by ByRefLike structs
- [x] reflection icall
- [x] Add `[PreserveDependencyAttribute (".ctor()", "System.Runtime.CompilerServices.IsByRefLikeAttribute")]` 
- [x] See https://github.com/mono/mono/issues/11220 - ~Run tests from https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Type/TypeTests.netcoreapp.cs#L208-L213~

### CODE REVIEW

**Comment by marek-safar** [comment]
@MichalStrehovsky could you help us here?

**Comment by lambdageek** [comment]
Is it just "has `System.Runtime.CompilerServices.IsByRefLikeAttribute`"?

**Comment by MichalStrehovsky** [comment]
@lambdageek yup, that's correct. The C# 7 language compilers emit this attribute on `ref struct`.

The runtime does a bit more with it:
* It enforces that a byref-like type can only be used as a field type if the owning type of the field is also byref-like and if it's an instance field
* It enforces that byref-like types can never be boxed and cannot be elements of an array.

Byref-like types are things like `Span<T>` or `TypedReference`.

File: mcs/class/corlib/System/RuntimeTypeHandle.cs [hunk 1]
```diff
@@ -258,5 +258,7 @@ internal static bool IsSubclassOf (RuntimeType childType, RuntimeType baseType)
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
 		internal extern static bool is_subclass_of (IntPtr childType, IntPtr baseType);
 
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
```
Comment (marek-safar) [inline mcs/class/corlib/System/RuntimeTypeHandle.cs]: Please add `[PreserveDependencyAttribute (".ctor()", "System.Runtime.CompilerServices. IsByRefLikeAttribute"` for linker tracking

**Comment by lambdageek** [comment]
Okay, this is ready for review.  Runtime and JIT changes.

**Comment by lambdageek** [comment]
Resolved merge conflicts.

File: mono/metadata/class-init.c [hunk 2]
```diff
@@ -701,6 +716,30 @@ mono_generic_class_setup_parent (MonoClass *klass, MonoClass *gtd)
 	mono_loader_unlock ();
 }
 
+struct HasIsByrefLikeUD {
```
Comment (jaykrell) [inline mono/metadata/class-init.c]: nit:  typedef to HasIsByrefLikeUD w/o struct

File: mono/metadata/icall-def.h [hunk 3]
```diff
@@ -894,6 +894,7 @@ HANDLES(RTH_9, "HasInstantiation", ves_icall_RuntimeTypeHandle_HasInstantiation,
 HANDLES(RTH_20, "HasReferences", ves_icall_RuntimeTypeHandle_HasReferences, MonoBoolean, 1, (MonoReflectionType))
 HANDLES(RTH_10, "IsArray", ves_icall_RuntimeTypeHandle_IsArray, MonoBoolean, 1, (MonoReflectionType))
 HANDLES(RTH_11, "IsByRef", ves_icall_RuntimeTypeHandle_IsByRef, MonoBoolean, 1, (MonoReflectionType))
+HANDLES(RTH_21, "IsByRefLike", ves_icall_RuntimeTypeHandle_IsByRefLike, MonoBoolean, 1, (MonoReflectionType))
```
Comment (jaykrell) [inline mono/metadata/icall-def.h]: I thought it'd have to be `RTH_11a`, but if it works, ok.

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -3893,6 +3893,12 @@ mini_emit_box (MonoCompile *cfg, MonoInst *val, MonoClass *klass, int context_us
 {
 	MonoInst *alloc, *ins;
 
+	if (G_UNLIKELY (m_class_is_byreflike (klass))) {
+		mono_error_set_bad_image (&cfg->error, m_class_get_image (cfg->method->klass), "Cannot box IsByRefLike type '%s.%s'", m_class_get_name_space (klass), m_class_get_name (klass));
+		mono_cfg_set_exception (cfg, MONO_EXCEPTION_MONO_ERROR);
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: Tests for all the error paths?

File: mono/metadata/icall-def.h (hunk 3 above)
Comment (marek-safar) [reply mono/metadata/icall-def.h]: This requires mscorlib version bump (otherwise LGTM)

**Comment by lambdageek** [comment]
Ok, I'm going to add the version bump and try to add some IL negative tests, if possible.

**Comment by lambdageek** [comment]
Hm. tests are good.  - I expected the IL equivalent of`G<R> o = new G<R>()` where `R` is a `ref struct` to throw some kind of exception, but it didn't.  Investigating...

File: mono/metadata/custom-attrs.c [hunk 5]
```diff
@@ -2397,6 +2411,23 @@ mono_assembly_metadata_foreach_custom_attr (MonoAssembly *assembly, MonoAssembly
...
 
+	metadata_foreach_custom_attr_from_index (image, idx, func, user_data);
+}
+
+/**
+ * iterate over the custom attributes that belong to the given index and call func, passing the 
+ *  assembly ref (if any) and the namespace and name of the custom attribute.
+ *
+ * Everything is done using low-level metadata APIs, so it is fafe to use
+ * during assembly loading and class initialization.
```
Comment (vargaz) [inline mono/metadata/custom-attrs.c]: Typo.

**Comment by lambdageek** [comment]
Now with some IL tests.

Had to add a new check to `mono_class_create_generic_inst` to fail classes that are invalid instantiations.

The issue is that in the metadata parser we create a `MonoGenericContext` separately from the generic type definition that it will be attached to, and then bring them together right in the parser without going through `mono_class_inflate_generic_type_checked`. 

This happens for IL locals that are generic instances, we basically create those `MonoType`s without going through the inflation machinery and consequently miss sanity-checking the instantiation.

So now we'll do it when we actually try to make a `MonoClass` out of the `MonoType`.

(Note: checking in the parser didn't work because it broke an existing test.  we can't be too eager to fail.)

**Comment by lambdageek** [comment]
@monojenkins build deb with monolite

**Pull request merged**